service/storage_service.cc

```diff
@@ -681,44 +694,76 @@ future<> storage_service::merge_topology_snapshot(raft_topology_snapshot snp) {
+    on_endpoint_change(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, const char* ev, gms::permit_id permit_id) {
         auto app_state_ptr = ep_state->get_application_state_ptr(gms::application_state::HOST_ID);
         if (!app_state_ptr) {
             co_return;
         }
         raft::server_id id(utils::UUID(app_state_ptr->value()));
-        rslog.debug("raft_ip_address_updater::on_endpoint_change() {} {}", endpoint, id);
+        rslog.debug("raft_ip_address_updater::on_endpoint_change({}) {} {}", ev, endpoint, id);
 
         const auto prev_ip = _address_map.find(id);
         _address_map.add_or_update_entry(id, endpoint, ep_state->get_heart_beat_state().get_generation());
+        if (prev_ip == endpoint) {
+            co_return;
+        }
+        if (_address_map.find(id) == prev_ip) {
+            // Address map refused to update IP for the host_id,
+            // this means prev_ip has higher generation than endpoint.
+            // We can immediately remove endpoint from gossiper
+            // since it represents and old IP (before an IP change)
+            // for the given host_id. This is not strictly
+            // necessary, but it reduces the noice circulated
+            // in gossiper messages and allows for clearer
+            // expectations of the gossiper state in tests.
+
+            co_await _ss._gossiper.force_remove_endpoint(endpoint, permit_id);
+            co_return;
+        }
 
         // If the host_id <-> IP mapping has changed, we need to update system tables, token_metadat and erm.
-        if (_ss.raft_topology_change_enabled() && prev_ip != endpoint && _address_map.find(id) == endpoint) {
-            rslog.debug("raft_ip_address_updater::on_endpoint_change(), host_id {}, "
+        if (_ss.raft_topology_change_enabled()) {
+            rslog.debug("raft_ip_address_updater::on_endpoint_change({}), host_id {}, "
                         "ip changed from [{}] to [{}], "
                         "waiting for group 0 read/apply mutex before reloading Raft topology state...",
-                id, prev_ip, endpoint);
+                ev, id, prev_ip, endpoint);
 
             // We're in a gossiper event handler, so gossiper is currently holding a lock
             // for the endpoint parameter of on_endpoint_change.
             // The topology_state_load function can also try to acquire gossiper locks.
             // If we call sync_raft_topology_nodes here directly, a gossiper lock and
             // the _group0.read_apply_mutex could be taken in cross-order leading to a deadlock.
             // To avoid this, we don't wait for sync_raft_topology_nodes to finish.
-            (void)_ss._group0->client().hold_read_apply_mutex().then([this, id, endpoint](semaphore_units<> g) {
+            (void)_ss._group0->client().hold_read_apply_mutex().then(ensure_alive([this, id, endpoint, prev_ip, h = _ss._async_gate.hold()](semaphore_units<> g) -> future<> {
```

Comment:
This problem was solved in existing code by turning the captures into arguments and then binding them with `bind_front`. See e.g. `raft_group0::discover_group0`
```cpp
    }).finally(std::bind_front([] (raft_group0& self, persistent_discovery& p_discovery) -> future<> {
        co_await p_discovery.stop();
        self._group0 = std::monostate{};
    }, std::ref(*this), std::ref(p_discovery)));
```

Reply:
Also why leave this single `then`? It could be put into the coroutine too.